File: api/square-occ.py
import gmsh
import sys

gmsh.initialize()
gmsh.model.add("square")

# define params
Lc = 1e-1
Lx = 1
Ly = 1

# create surface
s = gmsh.model.occ.addRectangle(-Lx, -Ly, 0, 2*Lx, 2*Ly)
print(f's={s}', flush=True)

# Force synchro
gmsh.model.occ.synchronize()

# Enumerate enties
for dim in range(3):
    print(f'entities(dim={dim})={ gmsh.model.getEntities(dim)}', flush=True)

    print(f'boundary: {gmsh.model.getBoundary([(2,s)])}', flush=True)
for (dim, tag) in gmsh.model.getBoundary([(2,s)]):
    otype = gmsh.model.getType(dim, abs(tag))
    print(f'dim: {dim}, tag:{tag}, otype: {otype}')

# Select entities with boundingbox
eps = 1.e-3
selected = gmsh.model.getEntitiesInBoundingBox(-Lx-eps, -Ly-eps, -eps, Lx+eps, -Ly+eps, eps, 1)
print(f'selected: {selected}')
ps = gmsh.model.addPhysicalGroup(1, [tag for (dim, tag) in selected])
gmsh.model.setPhysicalName(1, ps, "My Line")

# Physical Surface
ps = gmsh.model.addPhysicalGroup(2, [s])
gmsh.model.setPhysicalName(2, ps, "My surface")

gmsh.model.mesh.generate(2)
gmsh.write("square.msh")

if '-nopopup' not in sys.argv:
    gmsh.fltk.run()

File: cpp/square-occ.cpp
#include <set>
#include <gmsh.h>
int main(int argc, char **argv)
{
  gmsh::initialize();

  double lc = 0.1;
  gmsh::model::add("square");
  gmsh::model::geo::addPoint(.1, 0, 0, lc, 2);
  
  double roundedRadius = 0;
  int tag = -1;
  gmsh::model::occ::addRectangle(-0.5, -0.5, -0.5, 1, 1, tag, roundedRadius);

  gmsh::model::occ::synchronize();
  gmsh::model::addPhysicalGroup(2, {1}, -1, "My surface");

  std::set<std::string> args(argv, argv + argc);
  if(!args.count("-nopopup")) gmsh::fltk::run();
  
  gmsh::finalize();
  return 0;
}

